### PR TITLE
CLOUDP-67960: update internal/cli/atlas/backup to use the new output and fix point in time backup

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ CLOUDP-67960

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

## Further comments

**OUTPUT:**

```
bin/mongocli atlas backup snapshots list Cluster0 -o ""
ID				TYPE		STATUS		CREATED AT		EXPIRES AT
5f2173a686f2dd69cb169df3	onDemand	completed	2020-07-29T13:06:32Z	2020-08-01T13:07:32Z
5f21672423c74b4f5fb6ec29	onDemand	completed	2020-07-29T12:14:22Z	2020-08-01T12:15:26Z
5f21627f8f4f772aff4f3da3	onDemand	completed	2020-07-29T11:52:54Z	2020-08-01T11:53:53Z
```

```
bin/mongocli atlas backup restores list e2e-cluster-1563900813 -o ""
ID				SNAPSHOT			CLUSTER		TYPE		EXPIRES AT
5f2163e323c74b4f5fb6e9f1	5f21627f8f4f772aff4f3da3	Cluster0	automated	2020-07-29T15:56:19Z
```

```
bin/mongocli atlas backups restores start download --clusterName Cluster0 --snapshotId 5f201b515d6af95ecf27d8c9
Restore job '5f201b515d6af95ecf22d8c9' successfully started
```

```
bin/mongocli atlas backups restores start pointInTime --clusterName Cluster2 --targetClusterName Cluster3 --targetProjectId 5e4e593f70dfbf1010295836 --pointInTimeUTCMillis 1596020517 -o ""
Restore job '5f201b515d6af95ecf22d8c9' successfully started
```

```
bin/mongocli atlas backups restores start automated --clusterName Cluster4 --snapshotId 5f215b7c23c74b4f5fb6e5ed --targetClusterName Cluster4 --targetProjectId 5e4e593f70dfbf1010295836 -o ""
Restore job '5f201b515d6af95ecf22d8c9' successfully started
```

**JSON OUTPUT:**

```
bin/mongocli atlas backups restores start pointInTime --clusterName Cluster0 --targetClusterName Cluster1 --targetProjectId 5e4e593f70dfbf1010295836 --pointInTimeUTCMillis 1596008100
{
  "id": "5f2146c10edf6a7e25499e0f",
  "snapshotId": "5f2118445d529346d492e6a0",
  "deliveryType": "automated",
  "targetClusterName": "Cluster1",
  "targetGroupId": "5e4e593f70dfbf1010295836",
  "expiresAt": "2020-07-29T13:52:01Z",
  "links": [
    {
      "rel": "self",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/restoreJobs/5f2146c10edf6a7e25499e0f"
    },
    {
      "rel": "http://mms.mongodb.com/snapshot",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/snapshots/5f2118445d529346d492e6a0"
    },
    {
      "rel": "http://mms.mongodb.com/cluster",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0"
    },
    {
      "rel": "http://mms.mongodb.com/group",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836"
    }
  ],
  "timestamp": "2020-07-29T06:35:25Z"
}
```

```
bin/mongocli atlas backups restores start automated --clusterName Cluster0 --snapshotId 5f201b515d6af95ecf27d8c9 --targetClusterName Cluster1 --targetProjectId 5e4e593f70dfbf1010295836
{
  "id": "5f201e6d5d6af95ecf27dae4",
  "snapshotId": "5f201b515d6af95ecf27d8c9",
  "deliveryType": "automated",
  "targetClusterName": "Cluster1",
  "targetGroupId": "5e4e593f70dfbf1010295836",
  "expiresAt": "2020-07-28T16:47:41Z",
  "links": [
    {
      "rel": "self",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/restoreJobs/5f201e6d5d6af95ecf27dae4"
    },
    {
      "rel": "http://mms.mongodb.com/snapshot",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/snapshots/5f201b515d6af95ecf27d8c9"
    },
    {
      "rel": "http://mms.mongodb.com/cluster",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0"
    },
    {
      "rel": "http://mms.mongodb.com/group",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836"
    }
  ],
  "timestamp": "2020-07-28T12:36:27Z"
}
```

```
bin/mongocli atlas backups restores start download --clusterName Cluster0 --snapshotId 5f201b515d6af95ecf27d8c9
{
  "id": "5f214af48f4f772aff4f2c01",
  "snapshotId": "5f201b515d6af95ecf27d8c9",
  "deliveryType": "download",
  "expiresAt": "2020-07-29T14:09:56Z",
  "links": [
    {
      "rel": "self",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/restoreJobs/5f214af48f4f772aff4f2c01"
    },
    {
      "rel": "http://mms.mongodb.com/snapshot",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/snapshots/5f201b515d6af95ecf27d8c9"
    },
    {
      "rel": "http://mms.mongodb.com/cluster",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0"
    },
    {
      "rel": "http://mms.mongodb.com/group",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836"
    }
  ],
  "timestamp": "2020-07-28T12:36:27Z"
}
```

### CODE REVIEW

File: internal/cli/atlas/backup/restores_list.go [hunk 1]
```diff
@@ -38,14 +38,18 @@ func (opts *RestoresListOpts) initStore() error {
 	return err
 }
 
+var restoreListTemplate = `ID	SNAPSHOT	CLUSTER	TYPE	EXPIRE{{range .Results}}
```
Comment (gssbzn) [inline internal/cli/atlas/backup/restores_list.go:41]: [nit-grammar] missing an `S` but I'd keep the `AT` to indicate is a date
```suggestion
var restoreListTemplate = `ID	SNAPSHOT	CLUSTER	TYPE	EXPIRES AT{{range .Results}}
```

File: internal/cli/atlas/backup/restores_start.go [hunk 2]
```diff
@@ -63,25 +69,38 @@ func (opts *RestoresStartOpts) Run() error {
...
 	}
 
-	return output.Print(config.Default(), "", r)
+	if opts.isPointInTimeRestore() {
+		return output.Print(config.Default(), pointInTimeTemplate, r)
+	} else if opts.isAutomatedRestore() {
+		return output.Print(config.Default(), automatedTemplate, r)
+	}
+
+	return output.Print(config.Default(), downloadTemplate, r)
```
Comment (gssbzn) [inline internal/cli/atlas/backup/restores_start.go:78]: [nit] the only difference is the template, create a function `func (opts*) template() string` or something and just do
```suggestion
	return output.Print(config.Default(), opts.template(), r)
```

File: internal/cli/atlas/backup/restores_start.go [hunk 3]
```diff
@@ -55,6 +54,13 @@ func (opts *RestoresStartOpts) initStore() error {
 	return err
 }
 
+var automatedTemplate = "Restoring {{.TargetClusterName}} using snapshot {{.SnapshotID}}\n"
+var pointInTimeTemplate = "Restoring {{.TargetClusterName}} using point in time\n"
+var downloadTemplate = `Links to download the snapshot:
+REL	HREF{{range .Links}}
+{{.Rel}}	{{.Href}}{{end}}
```
Comment (gssbzn) [inline internal/cli/atlas/backup/restores_start.go:61]: According to the docs the download links come in `deliveryUrl`, `Links` is usually a reference to the resource or resources

File: internal/cli/atlas/backup/restores_start.go [hunk 4]
```diff
@@ -91,41 +110,58 @@ func (opts *RestoresStartOpts) isAutomatedRestore() bool {
...
+
 	return cmd.MarkFlagRequired(flag.ClusterName)
 }
 
+func markRequiredDownloadRestoreFlags(cmd *cobra.Command) error {
+	if err := cmd.MarkFlagRequired(flag.SnapshotID); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (gssbzn) [inline internal/cli/atlas/backup/restores_start.go:162]: [nit this is the same as]
```suggestion
	return cmd.MarkFlagRequired(flag.SnapshotID)
```
which brings the question, do you need this function?

File: internal/cli/atlas/backup/snapshots_list.go [hunk 5]
```diff
@@ -38,14 +38,18 @@ func (opts *SnapshotsListOpts) initStore() error {
 	return err
 }
 
+var listTemplate = `ID	TYPE	STATUS	CREATE	EXPIRE{{range .Results}}
```
Comment (gssbzn) [inline internal/cli/atlas/backup/snapshots_list.go:41]: [nit-grammar] similar to above, missing `D` and `S` but is standard for dates to include an `AT`
```suggestion
var listTemplate = `ID	TYPE	STATUS	CREATED AT	EXPIRES AT{{range .Results}}
```

File: internal/cli/atlas/backup/restores_start.go (hunk 3 above)
Comment (andreaangiolillo) [reply internal/cli/atlas/backup/restores_start.go:61]: yes, you are right. The problem is that I don't have that field in the response. The doc says 
```
NOTE

If empty, Atlas is processing the restore job. Use the Get All Cloud Backup Restore Jobs endpoint periodically check for a deliveryUrl download value for the restore job.
```
but my atlas is not restoring any job. 🤔

Comment (andreaangiolillo) [reply internal/cli/atlas/backup/restores_start.go:61]: Response:

```
{
  "id": "5f2175f5cad06326fabb828d",
  "snapshotId": "5f21627f8f4f772aff4f3da3",
  "deliveryType": "download",
  "expiresAt": "2020-07-29T17:13:25Z",
  "links": [
    {
      "rel": "self",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/restoreJobs/5f2175f5cad06326fabb828d"
    },
    {
      "rel": "http://mms.mongodb.com/snapshot",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/snapshots/5f21627f8f4f772aff4f3da3"
    },
    {
      "rel": "http://mms.mongodb.com/cluster",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0"
    },
    {
      "rel": "http://mms.mongodb.com/group",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836"
    }
  ],
  "timestamp": "2020-07-29T11:52:54Z"
}
```

Comment (andreaangiolillo) [reply internal/cli/atlas/backup/restores_start.go:61]: then If I run again the same mongocli command I get:

`Error: POST https://cloud-dev.mongodb.com/api/atlas/v1.0/groups/5e4e593f70dfbf1010295836/clusters/Cluster0/backup/restoreJobs: 400 (request "Bad Request") Cannot create a manual download of a snapshot when there is already an active download for that snapshot.`

Comment (andreaangiolillo) [reply internal/cli/atlas/backup/restores_start.go:61]: maybe here I can just change the template to `Manual download created for snapshot {{.SnapshotID}}`

Comment (andreaangiolillo) [reply internal/cli/atlas/backup/restores_start.go:61]: I think I know what is the problem here. As you can see from the picture below, atlas needs time to create the download link. So we are not going to have the download link in the response of this command.

I think we should update the template to `Manual download created for snapshot {{.SnapshotID}}`. 
Let me know what you think.

Screenshot 2020-07-29 at 15 25 33

File: internal/cli/atlas/backup/restores_start.go [hunk 6]
```diff
@@ -55,6 +54,20 @@ func (opts *RestoresStartOpts) initStore() error {
...
+
+func (opts *RestoresStartOpts) template() string {
+	if opts.isPointInTimeRestore() {
+		return pointInTimeTemplate
+	} else if opts.isAutomatedRestore() {
+		return automatedTemplate
+	}
+
+	return downloadTemplate
+}
```
Comment (gssbzn) [inline internal/cli/atlas/backup/restores_start.go:69]: Given the challenge we have we http restore jobs I'd say we just show something
```suggestion
var startTemplate = "Restore job '{{.ID}}' successfully started\n"
```
Users will need the ID do do a get later if they want to check the status

File: internal/cli/atlas/backup/restores_start.go (hunk 3 above)
Comment (gssbzn) [reply internal/cli/atlas/backup/restores_start.go]: I think given the circumstances I would simplify the template even more, we can just return the ID of the JOB and let them query it with list or describe

**Pull request merged**